sympy/solvers/diophantine/diophantine.py

```diff
@@ -51,6 +52,106 @@
     "univariate"}
 
 
+class DiophantineSolutionSet(set):
```

Comment:
Do we want to introduce a non-SymPy object or should we be using current containers (like FiniteSet) to do this?

Reply:
Should we think bigger in terms of simply calling this `SolutionSet` or `NamedElementTupleSet` or...? This `n x m` object is basically a collection of n-dimensional Tuples in m free variables.

Reply:
If this is only an internal class then I think it's fine.

Reply:
It's basically just a union of image sets:

```
    def rewrite_as_Set(self):
        from sympy.sets import ImageSet, Union
        from sympy.core import Lambda

        image_sets = []
        base_set = [S.Integers]*len(self.parameters)

        for solution in self:
            image_sets.append(ImageSet(Lambda(self.parameters, solution), *base_set))

        return Union(*image_sets)
```

The advantage of not just using `Set` classes is that this is much easier to work with and reason about.

And there is the assumption that `_diop_linear` etc are internal functions.